### ISSUE TITLE
Getting Help (and getting commands)

### ISSUE THREAD

**Issue by Andrew74L** [PowerShell/PowerShell#19076]
### Summary of the new feature / enhancement

Some ideas for improving Get-Help.

1) Get-Help should accept pipeline input.
This:
```
Verb-Noun | Get-Help
Verb-Noun _defaultParameterValue_
```
takes a couple more keystrokes than:
```
Get-Help Verb-Noun
Verb-Noun _defaultParameterValue_
```
However, this:
`Verb-Noun | Get-Help -Detailed`
is preferable to:
`Get-Help Verb-Noun -Detailed`
Partly because the cursor goes to the right (correct & side) end of the previous command line when user depresses up arrow.

2) Get-Help should have closer integration with parameter names.
Instead of:
`Get-Help Verb-Noun -Parameter _parametername_`
this should work:
`Verb-Noun -_ParameterName_ | Get-Help `

3) Get-Help should have a paging switch parameter.
Instead of:
`Get-Help Verb-Noun -Detailed | Out-Host -Paging`
this should be possible:
`Verb-Noun | Get-Help -Detailed -Paging`

4) `Verb-Noun -?` is equivalent to `Get-Help Verb-Noun`. Great, but why stop there?
This should be possible:
`Verb-Noun [-Examples | -Detailed | -Full | -Online]`
That is, why not add these switches to common parameters?

5) Get-Help output would be easier to read with some coloring, especially examples.
Copy & paste any help out output into VSCode. That might be going to far, but some $PSStyle would be appreciated.

6) Get-Command should also accept pipeline input.
This should be possible:
`Verb-Noun | Get-Command -Module`

7) Get-Help remarks should include module information.
```
REMARKS
    Verb-Noun comes from the _ModuleName_ module.
    To see all the commands in this module, type: "Verb-Noun | Get-Command -Module"
    Alternatively, type: "Get-Command -Module _ModuleName_"
```
I think this would aid the discoverability of commands.

**Comment by 237dmitry** [comment]
> Partly because the cursor goes to the right (correct & side) end of the previous command line when user depresses up arrow.

User can press `upArrow; F1` for detailed help or type `Space;-?` for brief one.

**Comment by mklement0** [comment]
Indeed, the `PSReadLine` module offers convenient functions for invoking help context-sensitively.

The default key bindings are:

```powershell
PS> Get-PSReadLineKeyHandler | oss | Select-String help

F1            ShowCommandHelp   Shows help for the command at the cursor in an alternate screen buffer.
Alt+h         ShowParameterHelp Shows help for the parameter at the cursor.
```

As stated, `ShowCommandHelp`  uses an _alternate screen buffer_ (a separate screen, so to speak) to show `Get-Help -Full` output for the command to the left of the cursor _with paging_; if there's a _parameter_ to the left of the cursor, the relevant description inside the command's help topic is directly jumped to. On typing `q`, the original screen is restored.
  * Note: 
    * As of this writing, the parameter name must be spelled out in full (use tab-completion); while not a hardship, the behavior is surprising and it would be nice if that weren't required - see https://github.com/PowerShell/PSReadLine/issues/3560
    * Parameter _aliases_ aren't supported (e.g., `-lp` for `-LiteralPath`), which will hopefully be fixed - see #13418
    * On macOS, the use of `Alt`-based keyboard shortcuts requires one-time configuration of the terminal application - see https://github.com/PowerShell/PSReadLine/issues/1075#issuecomment-1412449026

`ShowParameterHelp` shows type information and the description for the _parameter_ to the left of the cursor _below the current command line_ (the same constraints re recognizing parameter names as above apply).

**Comment by mklement0** [comment]
The above I think addresses many of your usability concerns.

To address your suggestions: `Get-Help` _does_ accept pipeline input, but only _via an object whose `.Name` property contains the command name_:

```
[pscustomobject] @{ Name = 'Get-Date' }  | Get-Help
```

This is somewhat helpful when piping the output of a `Get-Command` call:

```powershell
Get-Command Get-Date | Get-Help
```

However, that is more verbose than `Get-Help Get-Date` or - with paging, using the built-in `help` function - `help Get-Date`.

The pipeline is not a good fit here in general, as only _one_ topic can be meaningfully processed.

More fundamentally, while something like `'Get-Date' | Get-Help` could be made to work (though doesn't seem worth it, the following suggested by you can _not_ work (the same applies to your `Get-Command` suggestion):

```powershell
# !! Cannot work syntactically, because Get-Date is *invoked*.
Get-Date | Get-Help

# !! Ditto
Get-Date -Year | Get-Help
```

> This should be possible: `Verb-Noun [-Examples | -Detailed | -Full | -Online]`

Providing more control over what help content is displayed without involving `Get-Help` would be nice, but we don't want to introduce new common parameters unless absolutely necessary, given the invariable risk of name collisions (especially with switches such as `-Online`.

See https://github.com/PowerShell/PowerShell/issues/4137 for a proposed solution based one extending `-?`

As for the remaining suggestions, it's probably better to open separate, focused issues:

> Get-Help output would be easier to read with some coloring, especially examples.

Good idea. Not only would coloring help, but also a better formatting of the syntax diagrams, with richer information.
Several issues in that vein already exist:

* #8692
* #6856
* #4715
* #4135
* #4626

> Get-Help remarks should include module information.

Good idea.

**Comment by Andrew74L** [comment]
> > Partly because the cursor goes to the right (correct & side) end of the previous command line when user depresses up arrow.
> 
> User can press `upArrow; F1` for detailed help or press `upArrow` and type `Space-?` for brief one.

That may be true, but is it reasonable to expect that user knows about this?

A function called `Create-Shortcut` cannot be authored without getting a squiggly line in vscode, and a WARNING from `Import-Module`, because the unchecked proliferation of verbs, supposedly jeopardizes the discoverability of commands. How is the hypothetical user supposed to discover PSReadLine key sequences?
If it is deemed reasonable that key sequences are learned, then why is not also reasonable that the user learns about commands by typing; `(Get-Command Vern-Noun).Module.ExportedCommands`? Then we could do away with approved verbs, and start naming commands with more appropriate semantics. Ex: `End-Process` or `Kill-Process` vs `Stop-Process`.

My point is; what sort of user is Powershell usability aimed at? If the general answer is "system admins", then I would suppose that that user would prefer;
`End-Task` to `Stop-Process`, because the first is a closer match to the Windows GUI, and also prefer;
`Get-ADUser | Get-Help -Detailed` to `Get-Help Get-ADUser -Detailed`, because the first is closer to the familiar `some.exe /?` syntax, and less 'hacking' of the command line is required when the user presses up arrow.

**Comment by Andrew74L** [comment]
> More fundamentally, while something like `'Get-Date' | Get-Help` could be made to work (though doesn't seem worth it, the following suggested by you can _not_ work (the same applies to your `Get-Command` suggestion):
> 
> ```powershell
> # !! Cannot work syntactically, because Get-Date is *invoked*.
> Get-Date | Get-Help
> 
> # !! Ditto
> Get-Date -Year | Get-Help
> ```
In the case of `Get-Date`, I'm inclined to ask; so what if it is invoked? `Get-Help` would see a date object, and could apply `Get` (which I think is the default verb) to that object, to derive `Get-Date`. The command `date` is supposedly identical to `Get-Date`, although I noticed that `date -?` does not give the expected result.
For commands with mandatory parameters, could Powershell get the first mandatory parameter, and if the value is `$null`, and the next token on the command line is `|`, send a name parameter down the pipeline, and finalize processing that command? So this would then work:
`PS > Copy-Item | Get-Help`
This too:
```
PS > Copy-Item | Out-Host
Copy-Item
```
Along the same line, this would also work:
`Copy-Item -ToSession | Get-Help`

Thanks for all the links!

**Comment by mklement0** [comment]
> That may be true, but is it reasonable to expect that user knows about this?

Let's try to fix that: 
* https://github.com/MicrosoftDocs/PowerShell-Docs/issues/9796

**Comment by mklement0** [comment]
> In the case of Get-Date, I'm inclined to ask; so what if it is invoked? 

What if the cmdlet does something destructive (which it obviously shouldn't?
What if it prompts for missing parameters?
What if the parameter is a switch, which doesn't need an argument?
...

(And there's no way to infer from a given _output object's type_ what cmdlet it came from.)

Creating syntax exceptions the way you propose is asking for trouble and conceptual confusion - I strongly advise against it.

Again, <kbd>F1</kbd> and <kbd>Alt-h</kbd> are the solution, and they just need to be documented better (see previous comment).

**Comment by Andrew74L** [comment]
> Creating syntax exceptions the way you propose is asking for trouble and conceptual confusion - I strongly advise against it.

How pleasing or confusing this would be to use, should be determined in a usability lab, and/or with an experimental feature.

> Again, F1 and Alt-h are the solution, and they just need to be documented better (see previous comment).

The fact that these keyboard shortcuts exist, indicates that the command line alternatives are inefficient.
`Alt-H` is okay, but `F1` is not fun to use - no PgUp|PgDn, spacebar scroll or scroll wheel, and 'Q' to escape is a bit odd.

**Comment by mklement0** [comment]
> How pleasing or confusing this would be to use, should be determined in a usability lab, and/or with an experimental feature.

Not every idea is worth exploring further, if there are fundamental reasons that suggest that it is a dead end.
Of course, deeming this to be the case in _this_ instance is my personal assessment, but in the absence of arguments to the contrary my original rationale stands. Let's see if others weigh in.

> Alt-H is okay, but F1 is not fun to use - no PgUp|PgDn, spacebar scroll or scroll wheel, and 'Q' to escape is a bit odd.

See the previously linked to:
*  https://github.com/PowerShell/PSReadLine/issues/3561

**Comment by Andrew74L** [comment]
So going back to your original rationale...

> What if the cmdlet does something destructive (which it obviously shouldn't?
> What if it prompts for missing parameters?
> What if the parameter is a switch, which doesn't need an argument? ...
> ...
> (And there's no way to infer from a given _output object's type_ what cmdlet it came from.)

Don't these arguments apply to `Verb-Noun -?`?
How is that syntax able to work without issue, that could not also work with `Verb-Noun | Get-Help`?
I would presume that as the Powershell engine is always in control of all operations, there are no fundamental obstacles to any syntax format.

> Again, F1 and Alt-h are the solution, and they just need to be documented better (see previous comment).

Those suggestions would be great if implemented, but what if the user wants `-Examples` or `-Online` help? Do they revert to using the command line?

**Comment by mklement0** [comment]
> Those suggestions would be great if implemented, but what if the user wants -Examples or -Online help? Do they revert to using the command line?

That would make for great additions to https://github.com/PowerShell/PSReadLine/issues/3561 - I've just updated the proposal there.

> Don't these arguments apply to `Verb-Noun -?`?

No; you can think of `-?` as a [common parameter](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_CommonParameters), like `-ErrorAction` (which is also why it only works for cmdlets and _advanced_ functions/scripts), and the invocation syntax clearly indicates the intent to seek help.

The bottom line is this: a given invocation of a command (say, `Get-ChildItem -File`) should always act the same, whether stand-alone or in a pipeline, and what specific command is in the _next_ pipeline segment should never fundamentally change the behavior of such an invocation.

**Comment by Andrew74L** [comment]
> > Don't these arguments apply to `Verb-Noun -?`?
> 
> No; you can think of `-?` as a [common parameter](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_CommonParameters), like `-ErrorAction` (which is also why it only works for cmdlets and _advanced_ functions/scripts), and the invocation syntax clearly indicates the intent to seek help.

Unlike `-ErrorAction` though, `-?` is not included in the tab-completion list, so it would be better to think of `-?` as a _hidden_ common parameter. Perhaps it should be unhidden and named `-Help`.

> The bottom line is this: a given invocation of a command (say, `Get-ChildItem -File`) should always act the same, whether stand-alone or in a pipeline, and what specific command is in the _next_ pipeline segment should never fundamentally change the behavior of such an invocation.

So your bottom line is that there should be no special cases. That leaves us with the following situation.
```
Verb-Noun -?  # for help
Get-Help Verb-Noun -Detailed  # for detailed help
Verb-Noun[F1]  # for full help
Get-Help Verb-Noun -Online  # for online help
Get-Help Verb-Noun -Parameter spelloutparameter  # for parameter help
Get-Help about_topic  # for language construct commands and other subjects help
Get-Command Verb-Noun -Syntax  # for syntax help
```
Compare all this with...
`anyCommand [Parameter] | Get-Help [-Detailed | -Full | -Online | -Syntax]`
Which is the least confusing?

**Comment by mklement0** [comment]
> `-?` is not included in the tab-completion list

Excellent point: please open a new issue focused on that.

> Which is the least confusing?

That 's easy; the following is the least confusing:

> a given invocation of a command (say, `Get-ChildItem -File`) should always act the same, whether stand-alone or in a pipeline, and what specific command is in the next pipeline segment should never fundamentally change the behavior of such an invocation.

As for `Get-Help about_topic` and the desire to get help for _language keywords_ such as `while` and `foreach`:

I agree that these should be easier to discover, and <kbd>F1</kbd> could be extended to recognize those as well.

Please open a feature suggestion at https://github.com/PowerShell/PSReadLine/issues (don't just comment on https://github.com/PowerShell/PSReadLine/issues/3561)

**Comment by theJasonHelmick** [comment]
@Andrew74L  Thank you for the issue -- The working group finds nothing actionable here. @mklement0 has provided answers for these issues.
